packages/permission-log-controller/src/PermissionLog.test.ts

```diff
@@ -0,0 +1,792 @@
+import { ControllerMessenger } from '@metamask/base-controller';
+import type {
+  JsonRpcEngineReturnHandler,
+  JsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import type { PendingJsonRpcResponse } from '@metamask/utils';
+import { nanoid } from 'nanoid';
+import { useFakeTimers } from 'sinon';
+import type { SinonFakeTimers } from 'sinon';
+
+import { LOG_LIMIT, LOG_METHOD_TYPES } from './enums';
+import { PermissionLogController } from './PermissionLog';
+import type {
+  JsonRpcRequestWithOrigin,
+  Permission,
+  PermissionActivityLog,
+} from './PermissionLog';
+import { CaveatTypes, constants, getters, noop } from './permissions';
+
+const { PERMS, RPC_REQUESTS } = getters;
+const {
+  ACCOUNTS,
+  EXPECTED_HISTORIES,
+  SUBJECTS,
+  PERM_NAMES,
+  REQUEST_IDS,
+  RESTRICTED_METHODS,
+} = constants;
+
+let clock: SinonFakeTimers;
+
+type LogMiddleware = (
+  req: JsonRpcRequestWithOrigin,
+  res: PendingJsonRpcResponse<(string | Permission)[]>,
+  next?: JsonRpcEngineNextCallback,
+) => void;
+
+const name = 'PermissionLogController';
+
+/**
+ * Constructs a restricted controller messenger.
+ *
+ * @returns A restricted controller messenger.
+ */
+function getMessenger() {
+  return new ControllerMessenger().getRestricted<typeof name, never, never>({
+    name,
+  });
+}
+
+const initPermLog = (state = {}): PermissionLogController => {
+  const messenger = getMessenger();
+  return new PermissionLogController({
+    messenger,
+    restrictedMethods: RESTRICTED_METHODS,
+    state,
+  });
+};
+
+const mockNext: JsonRpcEngineNextCallback = (handler) => {
+  if (handler) {
+    handler(noop);
+  }
+};
+
+const initMiddleware = (permLog: PermissionLogController): LogMiddleware => {
+  const middleware = permLog.createMiddleware();
+  return (
+    req: JsonRpcRequestWithOrigin,
+    res: PendingJsonRpcResponse<(string | Permission)[]>,
+    next = mockNext,
+  ) => {
+    middleware(req, res, next);
+  };
+};
+
+const initClock = () => {
+  // useFakeTimers, is in fact, not a react-hook
+  // eslint-disable-next-line
+  clock = useFakeTimers(1);
+};
+
+const tearDownClock = () => {
+  clock.restore();
+};
+
+const getSavedMockNext =
+  (
+    arr: (JsonRpcEngineReturnHandler | undefined)[],
+  ): JsonRpcEngineNextCallback =>
+  (handler) => {
+    arr.push(handler);
+  };
+
+describe('PermissionLogController', () => {
+  describe('restricted method activity log', () => {
+    let permLog: PermissionLogController, logMiddleware: LogMiddleware;
+
+    beforeEach(() => {
+      permLog = initPermLog();
+      logMiddleware = initMiddleware(permLog);
+    });
+
+    it('records activity for restricted methods', () => {
+      let log: PermissionActivityLog[],
+        req: JsonRpcRequestWithOrigin,
+        res: PendingJsonRpcResponse<(string | Permission)[]>;
+
+      // test_method, success
+
+      req = RPC_REQUESTS.test_method(SUBJECTS.a.origin);
+      req.id = REQUEST_IDS.a;
+      res = {
+        id: null,
```

Comment:
I see that these properties were added to this response object, as well as other response objects in this file. While these additional properties do match the `PendingJsonRpcResponse` type, I think we should keep these objects as they were to ensure that the middleware works the same way as it does now. (I am okay with liberal uses of `// @ts-expect-error` throughout this file to get TypeScript to play nice.) This also makes it easier to compare this version of the file with the JavaScript version.

What do you think?